tc/autotuner/autotuner-inl.h

```diff
@@ -0,0 +1,540 @@
+  CHECK(executors_.empty());
+  CHECK(configurations_.empty());
+
+  {
+    // Initialize for this round
+    currentCompilationJob_.store(0);
+    numEvaluations_.store(0);
+    Printer printer(
+        iteration,
+        searchStrategy.population.size(),
+        currentCompilationJob_,
+        numEvaluations_);
+    auto logIterations = FLAGS_tuner_gen_log_generations;
+    ScopeGuard sgPrinter([logIterations, &printer]() {
+      printer.stop();
+      if (logIterations) {
+        printer.printAll();
+      }
+    });
+
+    // Just spawn and join new threads for each iteration
+    std::vector<std::thread> cpuCompilationThreads;
+    cpuCompilationThreads.reserve(FLAGS_tuner_threads);
+    ScopeGuard sgCompilationThreads([&cpuCompilationThreads]() {
+      for (auto& cpuCompilationThread : cpuCompilationThreads) {
+        cpuCompilationThread.join();
+      }
+    });
+    for (size_t i = 0; i < FLAGS_tuner_threads; ++i) {
+      cpuCompilationThreads.emplace_back(
+          [this, &searchStrategy]() { this->doCompile(searchStrategy); });
+    }
+
+    // Just spawn and join new threads for each iteration
+    std::vector<std::thread> workerThreads;
+    workerThreads.reserve(devices.size());
+    LOG_IF(INFO, tc::FLAGS_debug_tuner)
+        << "Start evaluation: " << devices.size() << " " << executors_.size()
+        << " " << configurations_.size();
+    ScopeGuard sgDeviceWorkerThreads([&workerThreads]() {
+      for (auto& workerThread : workerThreads) {
+        workerThread.join();
+      }
+    });
+    auto populationSize = searchStrategy.population.size();
+    for (auto device : devices) {
+      workerThreads.emplace_back([this, device, populationSize, &printer]() {
+        this->doEvaluate(device, populationSize, printer);
+      });
+    }
+  }
+
+  // At this point everything is synchronized because out of scope, done
+  if (FLAGS_debug_tuner || FLAGS_tuner_print_best) {
+    LOG(INFO) << "[TUNER][ITERATION LOG] best option so far:";
+    std::stringstream ssInfo;
+    typename Backend::MappingOptionsCppPrinter infoPrinter(ssInfo);
+    infoPrinter << bestMappingOptions();
+    LOG_LINE_BY_LINE(INFO, ssInfo);
+  }
+  searchStrategy.updateParameters();
+}
+} // namespace detail
+
+namespace {
+volatile std::sig_atomic_t sigint_ = 0;
+volatile std::sig_atomic_t sigterm_ = 0;
+
+template <typename Backend>
+std::vector<typename Backend::MappingOptionsType> loadThroughCache(
+    lang::TreeRef tree,
+    std::shared_ptr<OptionsCache<Backend>> optionsCache,
+    const std::string& cacheFileName,
+    const std::vector<const DLConstTensor*>& inputs,
+    const size_t numCandidates) {
+  LOG_IF(INFO, FLAGS_debug_tuner)
+      << "Loading proto from: " << tc::makeOptionsFilename(cacheFileName)
+      << std::endl;
+  if (!cacheFileName.empty()) {
+    optionsCache->loadCacheFromFile(tc::makeOptionsFilename(cacheFileName));
+  }
+  tc::FLAGS_tuner_gen_restore_number =
```

Comment:
Sorry, I may have missed this part in the first pass (or it was not there).

Flags are essentially global variables.  Here, you change the global variable only to pass the result to a single function call, does not sound like a good idea.  What would happen if somebody wants to tune two different kernels in the same program (which does happen, e.g., in python) ?

Reply:
legacy.. not only what you mention but it is utterly useless, killing.